### PR TITLE
Add bitmap library and apply bitmap in fd_table

### LINKED ISSUE #6136: [RFC] Apply bitmap in fdTable
**Description**

Apply bitmap in fd_table to record open file FD. It can accelerate the speed of allocating FD for fdtable.
The time complexity to find an available FD is O(1) in the bitmap.

**Do you have a specific solution in mind?**

https://github.com/google/gvisor/pull/6257
1. add bitmap library. 
2. add fdBitmap in struct FDTable
3. changed related functions based on the fdBitmap

**Current tests status:**
- [x] unit-tests
- [x] nogo-tests
- [x] runsc-tests
- [x] syscall-tests kvm
- [x] syscall-tests ptrace

**Performance analysis**
Here is the benchmark for this PR.
1. time of creating 10000 socket

| original | bitmap | runc |
| --- | --- | --- |
| 1.076s | 0.700s | 0.070s |

2. time of some syscall tests (run each syscall tests five times and calculate the average time)

|| original | bitmap |
| --- | --- | --- |
|ping_socket_test_runsc_kvm_vfs2| 31.8s | 26.5s |
|socket_inet_loopback_nogotsan_test_runsc_kvm_vfs2| 51.2s | 44.5s |
|socket_ipv4_udp_unbound_loopback_nogotsan_test_runsc_kvm_vfs2| 26.9s | 18.1s |

### PR DESCRIPTION
**Pull request by zhlhahaha**
Provides the implementation of bitmap.
Apply bitmap in fd_table to record open file FD. It can accelerate the speed of allocating FD for fdtable.
The time complexity to find an available FD is O(1) in the bitmap.

Fixes #6136

**Current tests status:**
- [x] unit-tests
- [x] nogo-tests
- [x] runsc-tests
- [x] syscall-tests kvm
- [x] syscall-tests ptrace

Here is the benchmark for this PR.
1. time of create 10000 socket

| original | bitmap | runc |
| --- | --- | --- |
| 1.076s | 0.700s | 0.070s |

2. time of some syscall tests (run each syscall tests five times and calculate the average time)

|| original | bitmap |
| --- | --- | --- |
|ping_socket_test_runsc_kvm_vfs2| 31.8s | 26.5s |
|socket_inet_loopback_nogotsan_test_runsc_kvm_vfs2| 51.2s | 44.5s |
|socket_ipv4_udp_unbound_loopback_nogotsan_test_runsc_kvm_vfs2| 26.9s | 18.1s |

Signed-off-by: Howard Zhang <howard.zhang@arm.com>

### CODE REVIEW

**Comment by zhlhahaha** [comment]
Hi @ianlewis @kevinGC @ayushr2
I have resubmitted the PR here.

Do you have any suggestion on this? @ianlewis 
https://github.com/google/gvisor/pull/6208#discussion_r659841487

File: pkg/sentry/kernel/fd_table.go [hunk 1]
```diff
@@ -370,19 +409,14 @@ func (f *FDTable) NewFDsVFS2(ctx context.Context, fd int32, files []*vfs.FileDes
...
 	f.mu.Unlock()
 	return fds, nil
 }
 
-// NewFDVFS2 allocates a file descriptor greater than or equal to minfd for
+// NewFDVFS2 allocates a file descriptor greater than or equal to minFD for
 // the given file description. If it succeeds, it takes a reference on file.
-func (f *FDTable) NewFDVFS2(ctx context.Context, minfd int32, file *vfs.FileDescription, flags FDFlags) (int32, error) {
-	if minfd < 0 {
+func (f *FDTable) NewFDVFS2(ctx context.Context, minFD int32, file *vfs.FileDescription, flags FDFlags) (int32, error) {
```
Comment (ayushr2) [inline pkg/sentry/kernel/fd_table.go:418]: Hmm can't this just call `FDTable.NewFDsVFS2` with `files = []*vfs.FileDescription{file}`? The logic seems identical in both functions.

The point is that we should refactor this in some way to avoid duplicate logic.

File: pkg/sentry/kernel/fd_table.go [hunk 2]
```diff
@@ -396,29 +430,47 @@ func (f *FDTable) NewFDVFS2(ctx context.Context, minfd int32, file *vfs.FileDesc
...
-		if d, _, _ := f.getVFS2(fd); d == nil {
-			f.setVFS2(ctx, fd, file, flags)
-			if fd == f.next {
-				// Update next search start position.
-				f.next = fd + 1
-			}
-			return fd, nil
+	// max is the largest bit in fdBitmap
+	max := int32(0)
+	// flipFdBitmap is flip of fdBitmap which is used to find free fd in fdBitmap
```
Comment (ayushr2) [inline pkg/sentry/kernel/fd_table.go:443]: micro-nit: all comments should end with a full stop.

File: pkg/sentry/kernel/fd_table.go (hunk 1 above)
Comment (zhlhahaha) [reply pkg/sentry/kernel/fd_table.go:418]: > Hmm can't this just call `FDTable.NewFDsVFS2` with `files = []*vfs.FileDescription{file}`? The logic seems identical in both functions.
> 
> The point is that we should refactor this in some way to avoid duplicate logic.

 I think you are right. `NewFDsVFS2` can fulfill what `NewFDVFS2` can do.
Is it better to open a new PR for it?

Comment (ayushr2) [reply pkg/sentry/kernel/fd_table.go:418]: You can just update it here IMO.

**Comment by ayushr2** [comment]
Regarding https://github.com/google/gvisor/pull/6208#discussion_r659841487, I think the right thing to do is to use `uint32` in the `BitMap` implementation. In FDTable implementation, we should return EINVAL if negative FDs are used (I think we already do in many places). And we deal with the pain of type casts.

Note that some special FDs like `AT_FDCWD` are actually negative values and that is why the Linux syscall API uses signed integers for FDs. But I don't think any FD tracked by the FD table will be negative.

File: pkg/sentry/kernel/fd_table.go (hunk 1 above)
Comment (zhlhahaha) [reply pkg/sentry/kernel/fd_table.go:418]: ok

**Comment by zhlhahaha** [comment]
By the way, do you have any idea about how to add a state to fdBitmap, as I mentioned in the first comment?

**Comment by zhlhahaha** [comment]
> Regarding [#6208 (comment)](https://github.com/google/gvisor/pull/6208#discussion_r659841487), I think the right thing to do is to use `uint32` in the `BitMap` implementation. In FDTable implementation, we should return EINVAL if negative FDs are used (I think we already do in many places). And we deal with the pain of type casts.
> 
> Note that some special FDs like `AT_FDCWD` are actually negative values and that is why the Linux syscall API uses signed integers for FDs. But I don't think any FD tracked by the FD table will be negative.

Should we change the type of other parameters which are not supposed to be negative?
e.g. The begin and end in `func (b *Bitmap) flipRange(begin, end int)`

**Comment by ayushr2** [comment]
> Should we change the type of other parameters which are not supposed to be negative?
e.g. The begin and end in func (b *Bitmap) flipRange(begin, end int)

Yes I think that would make sense for begin and end to be uint32.

> do you have any idea about how to add a state to fdBitmap, as I mentioned in the first comment?

I think you have to add `+stateify savable` above the struct like [this](https://github.com/google/gvisor/blob/07ffecef83bd31e78786af901c49a7be93b20517/pkg/sentry/kernel/kernel.go#L92). And you can tag the fields that you don't want to save with \`state:"nosave"\`. We stateify library will autogenerate the necessary methods required to implement save and restore.

**Review by ayushr2 — APPROVED** [review]
Accepted with nits. Thanks for working on this.

File: pkg/sentry/kernel/fd_table.go [hunk 3]
```diff
@@ -573,7 +585,7 @@ func (f *FDTable) GetVFS2(fd int32) (*vfs.FileDescription, FDFlags) {
 // Precondition: The caller must be running on the task goroutine, or Task.mu
 // must be locked.
 func (f *FDTable) GetFDs(ctx context.Context) []int32 {
-	fds := make([]int32, 0, int(atomic.LoadInt32(&f.used)))
+	fds := make([]int32, 0, int(f.fdBitmap.GetNumOnes()))
```
Comment (kevinGC) [inline pkg/sentry/kernel/fd_table.go:588]: We'll need to lock before this since `f.fdBitmap` is protected by `f.mu`:
```go
f.mu.Lock()
defer f.mu.Unlock()
```

File: pkg/sentry/kernel/fd_table.go [hunk 4]
```diff
@@ -189,8 +188,10 @@ func (f *FDTable) DecRef(ctx context.Context) {
 func (f *FDTable) forEach(ctx context.Context, fn func(fd int32, file *fs.File, fileVFS2 *vfs.FileDescription, flags FDFlags)) {
```
Comment (kevinGC) [inline pkg/sentry/kernel/fd_table.go:188]: There now needs to be a lock/unlock in the call sites in saveDescriptorTable() and String()

File: pkg/sentry/kernel/fd_table.go [hunk 5]
```diff
@@ -590,6 +602,7 @@ func (f *FDTable) Fork(ctx context.Context) *FDTable {
 		if df, dfVFS2 := clone.setAll(ctx, fd, file, fileVFS2, flags); df != nil || dfVFS2 != nil {
 			panic("VFS1 or VFS2 files set")
 		}
+		clone.fdBitmap.Add(uint32(fd))
```
Comment (kevinGC) [inline pkg/sentry/kernel/fd_table.go:605]: Need to lock before the forEach.

File: pkg/sentry/kernel/fd_table_unsafe.go [hunk 6]
```diff
@@ -162,14 +165,6 @@ func (f *FDTable) setAll(ctx context.Context, fd int32, file *fs.File, fileVFS2
 		}
 	}
 
-	// Adjust used.
-	switch {
-	case orig == nil && desc != nil:
-		atomic.AddInt32(&f.used, 1)
-	case orig != nil && desc == nil:
-		atomic.AddInt32(&f.used, -1)
-	}
```
Comment (kevinGC) [inline pkg/sentry/kernel/fd_table_unsafe.go:171]: Why don't we need this anymore?

File: pkg/bitmap/bitmap.go [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+	b := new(Bitmap)
+	bSize := size / 64
+	if size % 64 != 0 {
+		bSize = bSize + 1
+	}
+	b.bitBlock = make([]uint64, bSize, bSize)
+	return b
+}
+
+// IsEmpty verifiys if the Bitmap is empty.
```
Comment (kevinGC) [inline pkg/bitmap/bitmap.go:45]: Nit/typo: "verifies whether"

File: pkg/bitmap/bitmap.go [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+// Clone the Bitmap.
+func (b *Bitmap) Clone() *Bitmap {
+	ptr := Bitmap{b.numOnes, make([]uint64, len(b.bitBlock))}
+	copy(ptr.bitBlock, b.bitBlock[:])
+	return &ptr
+}
+
+// ReplaceWith replace the content of b with that of newBitmap.
```
Comment (kevinGC) [inline pkg/bitmap/bitmap.go:105]: Is this called anywhere? If not, it can be deleted.

File: pkg/bitmap/bitmap.go [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+	b.numOnes = newBitmap.numOnes
+	b.bitBlock = make([]uint64, len(newBitmap.bitBlock))
+	copy(b.bitBlock, newBitmap.bitBlock[:])
+}
+
+// countOnesForBlocks count all 1 bits within b.bitBlock of begin and that of end.
+// The begin block and end block are inclusive.
+func (b *Bitmap) countOnesForBlocks(begin, end uint32) uint64 {
+	ones := uint64(0)
+	if begin > end {
```
Comment (kevinGC) [inline pkg/bitmap/bitmap.go:116]: This check can be removed.

File: pkg/bitmap/bitmap.go [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+func (b *Bitmap) flipRange(begin, end uint32) {
+	if begin > end {
+		return
+	}
+	if begin == end {
+		blockNum, mask := begin/64, uint64(1)<<(begin%64)
+		b.bitBlock[blockNum] = b.bitBlock[blockNum] ^ mask
+		return
+	}
+	beginBlock := begin / 64
```
Comment (kevinGC) [inline pkg/bitmap/bitmap.go:146]: You could probably delete everything in this function before this line.

File: pkg/bitmap/bitmap.go [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+func (b *Bitmap) clearRange(begin, end uint32) {
+	if begin > end {
+		return
+	}
+	if begin == end {
+		blockNum, mask := begin/64, uint64(1)<<(begin%64)
+		b.bitBlock[blockNum] = b.bitBlock[blockNum] &^ mask
+		return
+	}
+	beginBlock := begin / 64
```
Comment (kevinGC) [inline pkg/bitmap/bitmap.go:169]: You could probably delete everything in this function before this line.

File: pkg/bitmap/bitmap.go [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+// Bitmap struct contains numOnes which represent number of ones in the
+// bitmap and a slice bitBlock which represent bits. The type of bitBlock
+// is uint64 which means each number in bitBlock can contain information
+// for 64 bits.
+//
+// +stateify savable
+type Bitmap struct {
+	numOnes  uint32
+	bitBlock []uint64
+}
```
Comment (kevinGC) [inline pkg/bitmap/bitmap.go:32]: I think it makes more sense to put comments on the individual fields. Something like:
```go
// Bitmap implements an efficient bitmap.
//
// +stateify savable
type Bitmap struct {
	// numOnes is the number of ones in the bitmap.
	numOnes  uint32

	// bitBlock holds the bits. The type of bitBlock is uint64 which means
	// each number in bitBlock contains 64 entries.
	bitBlock []uint64
}
```

File: pkg/bitmap/bitmap.go [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+			// Interpret the bit as the in32 number it represents and add it to result.
+			bitmapSlice = append(bitmapSlice, uint32((base + int(bits.OnesCount64(j - 1)))))
+			bitBlock ^= j
+		}
+		base += 64
+	}
+	return bitmapSlice
+}
+
+// GetNumOnes return the numOnes of the Bitmap.
```
Comment (kevinGC) [inline pkg/bitmap/bitmap.go:234]: "GetNumOnes returns the number of ones in the Bitmap."

File: pkg/bitmap/bitmap.go [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+		b.numOnes = uint32(b.countOnesForAllBlocks())
+	} else {
+		oldRangeOnes := b.countOnesForBlocks(begin, end)
+		b.flipRange(begin, end)
+		newRangeOnes := b.countOnesForBlocks(begin, end)
+		b.numOnes += uint32(newRangeOnes - oldRangeOnes)
+	}
+}
+
+// ToSlice transform the Bitmap into slice.
```
Comment (kevinGC) [inline pkg/bitmap/bitmap.go:214]: // ToSlice transforms the Bitmap into a slice. For example, a bitmap of [0, 1, 0 , 1]
// will return the slice [1, 3].

File: pkg/bitmap/bitmap.go [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+// for 64 bits.
+//
+// +stateify savable
+type Bitmap struct {
+	numOnes  uint32
+	bitBlock []uint64
+}
+
+// NewBitmap create a new empty Bitmap.
+func NewBitmap(size uint32) *Bitmap {
```
[the remark this answers is not in the document]
Comment (zhlhahaha) [reply pkg/bitmap/bitmap.go:35]: Thanks, @kevinGC , I have some confusion here. 
> To avoid unnecessary allocation

Does this mean that avoid allocating data on the heap? 

> `func WithSize(size uint32) Bitmap {`

Is the function named with `WithSize` siffnix, like, `BitmapWithSize` or just named as `WithSize` ?

File: pkg/bitmap/bitmap.go (hunk 8 above)
Comment (zhlhahaha) [reply pkg/bitmap/bitmap.go:105]: This function was supposed to be used in `loadFdBitmap` which is state loader for fdBitmap in fd_table.go.
I will discuss this with you in the comment in `fd_table.go`

File: pkg/sentry/kernel/fd_table.go [hunk 16]
```diff
@@ -84,13 +84,8 @@ type FDTable struct {
...
 
-	// next is start position to find fd.
-	next int32
-
-	// used contains the number of non-nil entries. It must be accessed
-	// atomically. It may be read atomically without holding mu (but not
-	// written).
-	used int32
+	// fdBitmap shows which fds are already in use.
+	fdBitmap *bitmap.Bitmap `state:"nosave"`
```
[the remark this answers is not in the document]
Comment (zhlhahaha) [reply pkg/sentry/kernel/fd_table.go:88]: The fdBitmap would be restored in the function `loadDescriptorTable`. Does it still need to save it?

File: pkg/sentry/kernel/fd_table.go [hunk 17]
```diff
@@ -267,31 +267,56 @@ func (f *FDTable) NewFDs(ctx context.Context, fd int32, files []*fs.File, flags
...
-	if fd < f.next {
-		fd = f.next
+	// max is the largest bit in fdBitmap.
+	max := int32(0)
+	// flipFdBitmap is flip of fdBitmap which is used to find free fd in fdBitmap.
+	flipFdBitmap := f.fdBitmap.Clone()
+
+	if !f.fdBitmap.IsEmpty() {
+		max = int32(f.fdBitmap.Maximum())
+		flipFdBitmap.FlipRange(uint32(0), uint32(max))
```
[the remark this answers is not in the document]
Comment (zhlhahaha) [reply pkg/sentry/kernel/fd_table.go:284]: After moving the 2 above lines directly below the declaration of flipFdBitmap, the 0 bit would be flipped when fdBitmap is empty. This would lead to an error that FD 0 would be used even if the `minfd` is larger than 0. To fix this, here is a draft code.

```
@@ -278,16 +278,19 @@ func (f *FDTable) NewFDs(ctx context.Context, minFD int32, files []*fs.File, fla
        max := int32(0)
        // flipFdBitmap is flip of fdBitmap which is used to find free fd in fdBitmap.
        flipFdBitmap := f.fdBitmap.Clone()
+       max = int32(f.fdBitmap.Maximum())
+       flipFdBitmap.FlipRange(uint32(0), uint32(max))

        if !f.fdBitmap.IsEmpty() {
-               max = int32(f.fdBitmap.Maximum())
-               flipFdBitmap.FlipRange(uint32(0), uint32(max))
                max++
                // Clear 0 to (minFD-1) bit in flipFdBitmap, thus it will not use fd that is less than minFD.
                if minFD > 0 {
                       flipFdBitmap.ClearRange(uint32(0), uint32(minFD - 1))
                }
+       } else {
+               flipFdBitmap.ClearRange(uint32(0), uint32(0))
        }
+
        // Adjust max in case it is less than minFD.
        if max < minFD {
                max = minFD
@@ -296,7 +299,7 @@ func (f *FDTable) NewFDs(ctx context.Context, minFD int32, files []*fs.File, fla
        for len(fds) < len(files) {
                // Try to use free bit in fdBitmap.
                // If all bits in fdBitmap are used, expand fd to the max.
-               if !flipFdBitmap.IsEmpty() {
+               if !f.fdBitmap.IsEmpty() && !flipFdBitmap.IsEmpty() {
                        fd := flipFdBitmap.Minimum()
                        f.fdBitmap.Add(fd)
                        flipFdBitmap.Remove(fd)

```

File: pkg/sentry/kernel/fd_table.go (hunk 3 above)
Comment (zhlhahaha) [reply pkg/sentry/kernel/fd_table.go:588]: How about applying atomic in bitmap library?

File: pkg/sentry/kernel/fd_table_unsafe.go (hunk 6 above)
Comment (zhlhahaha) [reply pkg/sentry/kernel/fd_table_unsafe.go:171]: As f.used only used for count the total number of used fds in fd_table. `f.fdBitmap.GetNumOnes()` can achieve same goals.

File: pkg/bitmap/bitmap.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (kevinGC) [reply pkg/bitmap/bitmap.go:35]: Yes, it avoids allocating on the heap. And since there's no specific reason to use `*Bitmap`, `Bitmap` is simpler and faster.

And either `WithSize` or `BitmapWithSize` is fine. I was trying to avoid the "stutter" in `bitmap.BitmapWithSize`, but it's up to preference and I don't have a strong opinion.

File: pkg/sentry/kernel/fd_table.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (kevinGC) [reply pkg/sentry/kernel/fd_table.go:88]: Ah, I didn't realize that `loadDescriptorTable` was initializing the bitmap. In that case, `nosave` seems appropriate. But it still makes sense to change the type from `*bitmap.Bitmap` to `bitmap.Bitmap`.

File: pkg/sentry/kernel/fd_table.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (kevinGC) [reply pkg/sentry/kernel/fd_table.go:284]: Gotcha. Let's just leave it like it is then.

File: pkg/sentry/kernel/fd_table.go (hunk 3 above)
Comment (kevinGC) [reply pkg/sentry/kernel/fd_table.go:588]: I'm not sure how we could use an atomic here. We need the lock both to call `GetNumOnes` and to iterate over the list of FDs in `f.forEach`.

**Comment by kevinGC** [comment]
LGTM

**Comment by zhlhahaha** [comment]
Hi @kevinGC I am unable to see the failure in the `feedback/copybara` test, as the `Details` is an unreachable URL.

File: pkg/bitmap/bitmap.go [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+func (b *Bitmap) clearRange(begin, end uint32) {
+	end--
+	beginBlock := begin / 64
+	endBlock := end / 64
+	if beginBlock == endBlock {
+		b.bitBlock[beginBlock] &= (((uint64(1) << uint(begin%64)) - 1) | ^((uint64(1) << (uint(end) % 64 + 1)) - 1))
+	} else {
+		b.bitBlock[beginBlock] &= ((uint64(1) << uint(begin%64)) - 1)
+		for i := beginBlock + 1; i < endBlock; i++ {
+			b.bitBlock[i] &= ^b.bitBlock[i]
```
Comment (DarcySail) [inline pkg/bitmap/bitmap.go:150]: Why not directly assign a 0 here:
b.bitBlock[i] = 0

**Review by DarcySail — CHANGES_REQUESTED** [review]

**Comment by kevinGC** [comment]
Sorry @zhlhahaha. Let me try again and push on it more.

**Comment by avagin** [comment]
@zhlhahaha could you review https://github.com/avagin/gvisor/commit/fc72bab805e6af3fa2cd80e56826a38440dd09d2 and cherry pick it to this PR, if it looks good to you.

File: pkg/bitmap/bitmap.go (hunk 18 above)
Comment (zhlhahaha) [reply pkg/bitmap/bitmap.go:150]: > Why not directly assign a 0 here:
> b.bitBlock[i] = 0

Just thinking inertia, use bit set way to clean bitBlock.

**Comment by zhlhahaha** [comment]
> @zhlhahaha could you review [avagin@fc72bab](https://github.com/avagin/gvisor/commit/fc72bab805e6af3fa2cd80e56826a38440dd09d2) and cherry pick it to this PR, if it looks good to you.

Hi @avagin, the PR looks good to me, despite one place https://github.com/avagin/gvisor/commit/fc72bab805e6af3fa2cd80e56826a38440dd09d2#r54010072. 
I did a basic benchmark for this and it shows good performance improvement on Creating 10000 sockets.


| Test name | Before #6275 | On #6275 | On #6275 and #6368 |
| --- | --- | --- | --- |
| create 10000 socket | 0.932s | 0.547s | 0.392s |
| ping_socket_test_runsc_kvm_vfs2 | 30.2s | 21.7s | 21.8s |

**Comment by avagin** [comment]
I have added dup_benchmark and here are its results:

Before #6257:
```
---------------------------------------------------------------------------------
Benchmark                       Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------
BM_Dup/1/real_time           1688 ns         1685 ns       409580 items_per_second=592.42k/s
BM_Dup/8/real_time          11611 ns        11601 ns        62063 items_per_second=689.011k/s
BM_Dup/64/real_time        105401 ns       104636 ns         5543 items_per_second=607.206k/s
BM_Dup/512/real_time      1038014 ns      1038961 ns          693 items_per_second=493.249k/s
BM_Dup/4096/real_time    29317283 ns     29600000 ns           25 items_per_second=139.713k/s
BM_Dup/32768/real_time 1558471597 ns   1560000000 ns            1 items_per_second=21.0257k/s
```

On #6257:
```
---------------------------------------------------------------------------------
Benchmark                       Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------
BM_Dup/1/real_time          29060 ns        29039 ns        34092 items_per_second=34.4114k/s
BM_Dup/8/real_time         289503 ns       287873 ns         2779 items_per_second=27.6335k/s
BM_Dup/64/real_time       1996439 ns      1986607 ns          448 items_per_second=32.0571k/s
BM_Dup/512/real_time     17395519 ns     17555556 ns           45 items_per_second=29.4329k/s
BM_Dup/4096/real_time   116626999 ns    118000000 ns            5 items_per_second=35.1205k/s
BM_Dup/32768/real_time 1028151654 ns   1020000000 ns            1 items_per_second=31.8708k/s
```

On #6275 and #6368
```
---------------------------------------------------------------------------------
Benchmark                       Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------
BM_Dup/1/real_time           1925 ns         1923 ns       322337 items_per_second=519.459k/s
BM_Dup/8/real_time          15379 ns        15432 ns        45361 items_per_second=520.198k/s
BM_Dup/64/real_time        122617 ns       123880 ns         5247 items_per_second=521.951k/s
BM_Dup/512/real_time       968557 ns       965116 ns          860 items_per_second=528.621k/s
BM_Dup/4096/real_time     6801053 ns      6759259 ns          108 items_per_second=602.26k/s
BM_Dup/32768/real_time   68541736 ns     68461538 ns           13 items_per_second=478.074k/s
```

**Comment by zhlhahaha** [comment]
> @zhlhahaha could you review [avagin@fc72bab](https://github.com/avagin/gvisor/commit/fc72bab805e6af3fa2cd80e56826a38440dd09d2) and cherry pick it to this PR, if it looks good to you.

I have cherry pick the PR.

**Comment by avagin** [comment]
@zhlhahaha dup_test has to depend on @com_google_absl//absl/memory".  pls revert my previous commit and cherry pick https://github.com/avagin/gvisor/commit/3d6f04c60b855f7538e592bde7ebc0b436c4025a.

**Comment by zhlhahaha** [comment]
> @zhlhahaha dup_test has to depend on @com_google_absl//absl/memory". pls revert my previous commit and cherry pick [avagin@3d6f04c](https://github.com/avagin/gvisor/commit/3d6f04c60b855f7538e592bde7ebc0b436c4025a).

OK

**Comment by zhlhahaha** [comment]
> @zhlhahaha dup_test has to depend on @com_google_absl//absl/memory". pls revert my previous commit and cherry pick [avagin@3d6f04c](https://github.com/avagin/gvisor/commit/3d6f04c60b855f7538e592bde7ebc0b436c4025a).

It seems that [avagin@3d6f04c](https://github.com/avagin/gvisor/commit/3d6f04c60b855f7538e592bde7ebc0b436c4025a) does not contain correct CLAs.

**Comment by avagin** [comment]
You don't need to create a new branch:
$ git checkout mybranch
$ git fetch avagin bitmap
$ git cherry-pick avagin/bitmap

File: pkg/bitmap/bitmap.go [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+	// numOnes is the number of ones in the bitmap.
+	numOnes uint32
+
+	// bitBlock holds the bits. The type of bitBlock is uint64 which means
+	// each number in bitBlock contains 64 entries.
+	bitBlock []uint64
+}
+
+// BitmapWithSize create a new empty Bitmap.
+func BitmapWithSize(size uint32) Bitmap {
```
Comment (ayushr2) [inline pkg/bitmap/bitmap.go:36]: Our internal linters are catching this function name. It will be called `bitmap.BitmapWithSize` externally and the suggestion is to call is `bitmap.WithSize` instead.

I would say just rename this function to `NewBitmap` as this seems to be the only constructor. We would add `NewBitmapWithSize` to differentiate constructors.

**Review by ayushr2 — APPROVED** [review]

File: pkg/bitmap/bitmap.go (hunk 19 above)
Comment (avagin) [reply pkg/bitmap/bitmap.go:36]: I have renamed it to New.

**Comment by ayushr2** [comment]
@zhlhahaha thanks for working on this!
